lib/logstash/outputs/elasticsearch/http_client/pool.rb

```diff
@@ -282,19 +282,10 @@ def healthcheck!
               @logger.warn("Detected a node with a higher major version than previously observed. This could be the result of an elasticsearch cluster upgrade.", :previous_major => @maximum_seen_major_version, :new_major => major, :node_url => url.sanitized.to_s)
               set_new_major_version(major)
             end
-            if oss? || valid_es_license?(url)
-              meta[:state] = :alive
-            else
-              # As this version is to be shipped with Logstash 7.x we won't mark the connection as unlicensed
-              #
-              #  logger.error("Cannot connect to the Elasticsearch cluster configured in the Elasticsearch output. Logstash requires the default distribution of Elasticsearch. Please update to the default distribution of Elasticsearch for full access to all free features, or switch to the OSS distribution of Logstash.", :url => url.sanitized.to_s)
-              #  meta[:state] = :unlicensed
-              #
-              # Instead we'll log a deprecation warning and mark it as alive:
-              #
-              log_license_deprecation_warn(url)
-              meta[:state] = :alive
-            end
+
+            # license_check in mixed in from LogStash::Outputs::ElasticSearchPoolMixin::LicenseChecker
+            # separately defined in the elasticsearch output and the elasticsearch_data_streams output.
+            @license_checker.license_check!(url, meta)
```

Comment:
if we only had a pre-flight checker as I had suggested, that was given a reference to this pool and the `url` and simply returned true or false but _was not expected to manipulate pool internals_, then we could worry about pool internals here, in the pool implementation, and not in different consumers of this API in different code-bases.

```suggestion

            if @preflight.check?(this, url)
              meta[:state] = :alive
            else
              logger.warn("Connection failed pre-flight check", :url => url)
            end
```

The Elasticsearch Output then, could define its own pre-check that considered all of the stages in its license check without having to care about pool internals:

https://gist.github.com/yaauie/d73c4ab4284909cd3e0a5309ea3cd938#file-preflight_check-rb-L89